Implicit conversion never applied but influences type inference

## reproduction steps

https://stackoverflow.com/questions/62415172/mere-presence-of-implicit-conversion-makes-the-program-compile-despite-never-bei/62426914

using Scala 2.13.2,

```scala
scala> class Bar
class Bar

scala> def f[F[_], A](v: F[A]) = v
def f[F[_], A](v: F[A]): F[A]

scala> implicit def barToList(b: Bar): List[Int] = List(42)
def barToList(b: Bar): List[Int]

scala> f(new Bar)
val res1: Any = Bar@56881196
```

## problem

I would expect implicit conversion would result in

```
scala> f[List, Int](new Bar)
val res2: List[Int] = List(42)
```

where the following is inferred

```
F[_] = List
A = Int 
```

instead of actual 

```
F[_] = Any
A = Nothing 
```

Examining the output of `-Xprint:typer` shows the following implicit conversion never happend

```
f(barToList(new Bar()))
```

Dotty progressed. I think there is a ticket to improve `more specific than F[A]`.
```Scala
scala-cli compile --server=false -S 3 t12044.scala
-- [E007] Type Mismatch Error: /home/amarki/snips/t12044.scala:11:6 ------------
11 |    f(new Bar)
   |      ^^^^^^^
   |Found:    Bar
   |Required: F[A]
   |
   |where:    A is a type variable
   |          F is a type variable with constraint <: [_] =>> Any
   |
   |Note that implicit conversions were not tried because the result of an implicit conversion
   |must be more specific than F[A]
   |
   | longer explanation available when compiling with `-explain`
[[syntax trees at end of                     typer]] // /home/amarki/snips/t12044.scala
package <empty> {
  import language.implicitConversions
  class Bar() extends Object() {}
  final lazy module val Test: Test = new Test()
  final module class Test() extends Object(), App { this: Test.type =>
    def f[F[_ >: Nothing <: Any] >: Nothing <: Any, A >: Nothing <: Any](v: F[A]
      ): F[A] = v
    implicit def barToList(b: Bar): List[Int] = List.apply[Int]([42 : Int]*)
    Console.println(
      {
        Test.f[[_] =>> Any, Any](new Bar())
      }
    )
  }
}

1 error found
Compilation failed
```
With `-Yno-predef`, Scala 2 seems to be stuck.
```Scala
scala-cli compile --server=false -S 2 t12044.scala
/home/amarki/snips/t12044.scala:11: error: type mismatch;
 found   : Bar
 required: List[A]
    f(new Bar)
      ^
[[syntax trees at end of                     typer]] // t12044.scala
package <empty> {
  import scala.language.implicitConversions;
  class Bar extends scala.AnyRef {
    def <init>(): Bar = {
      Bar.super.<init>();
      ()
    }
  };
  object Test extends AnyRef with App {
    def <init>(): Test.type = {
      Test.super.<init>();
      ()
    };
    def f[F[_] >: [_]Nothing <: [_]Any, A](v: F[A]): F[A] = v;
    implicit def barToList(b: Bar): List[Int] = scala.`package`.List.apply[Int](42);
    scala.Console.println(f(new Bar()))
  }
}

1 error
Compilation failed
```

--- a/src/compiler/scala/tools/nsc/typechecker/Infer.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Infer.scala
@@ -15,9 +15,9 @@ package typechecker
 
 import scala.collection.{immutable, mutable}, mutable.ListBuffer
 import scala.reflect.internal.Depth
+import scala.tools.nsc.Reporting.WarningCategory, WarningCategory.LintInferAny
 import scala.util.control.ControlThrowable
 import symtab.Flags._
-import scala.tools.nsc.Reporting.WarningCategory
 
 /** This trait contains methods related to type parameter inference.
  *
@@ -55,7 +55,7 @@ trait Infer extends Checkable {
 
       val n = numArgs - numFormals + 1
       // Optimized version of: formals1.init ::: List.fill(n)(lastType)
-      val result = mutable.ListBuffer[Type]()
+      val result = ListBuffer.empty[Type]
       var fs = formals1
       while ((fs ne Nil) && (fs.tail ne Nil)) {
         result.addOne(fs.head)
@@ -495,10 +495,8 @@ trait Infer extends Checkable {
      *    type parameters that are inferred as `scala.Nothing` and that are not covariant in `restpe` are taken to be undetermined
      */
     def adjustTypeArgs(tparams: List[Symbol], tvars: List[TypeVar], targs: List[Type], restpe: Type = WildcardType): AdjustedTypeArgs  = {
-      val okParams = ListBuffer[Symbol]()
-      val okArgs = ListBuffer[Type]()
-      val undetParams = ListBuffer[Symbol]()
-      val allArgs = ListBuffer[Type]()
+      val okParams, undetParams = ListBuffer.empty[Symbol]
+      val okArgs, allArgs = ListBuffer.empty[Type]
 
       foreach3(tparams, tvars, targs) { (tparam, tvar, targ) =>
         val retract = (
@@ -586,11 +584,15 @@ trait Infer extends Checkable {
       // For example, don't require this arg, where the lub of `AnyRef` and `V` yields the `Any`.
       // this.forall(kv => map.getOrElse[Any](kv._1, Map.DefaultSentinelFn()) == kv._2)
       if (settings.warnInferAny && !fn.isEmpty)
-        foreach2(targs, tvars) { (targ, tvar) =>
+        foreach3(tparams, targs, tvars) { (tparam, targ, tvar) =>
           if (topTypes.contains(targ.typeSymbol) &&
               !tvar.constr.loBounds.exists(t => topTypes.contains(t.typeSymbol)) &&
               !tvar.constr.hiBounds.exists(t => topTypes.contains(t.typeSymbol)))
-            context.warning(fn.pos, s"a type was inferred to be `${targ.typeSymbol.name}`; this may indicate a programming error.", WarningCategory.LintInferAny)
+            context.warning(fn.pos, s"a type was inferred to be `${targ.typeSymbol.name
+              }`; this may indicate a programming error.", LintInferAny)
+          if (!tparam.isMonomorphicType && !targ.isHigherKinded)
+            context.warning(fn.pos, s"a type was inferred to be kind-polymorphic `${targ.typeSymbol.name
+              }` to conform to `${tparam.defString}`", LintInferAny)
         }
       adjustTypeArgs(tparams, tvars, targs, restpe)
     }

--- a/test/files/neg/t12044.check
+++ b/test/files/neg/t12044.check
@@ -1,0 +1,9 @@
+t12044.scala:51: warning: a type was inferred to be kind-polymorphic `Any` to conform to `F[_]`
+    f(new Bar) // warn
+    ^
+t12044.scala:54: warning: a type was inferred to be kind-polymorphic `Any` to conform to `F[_]`
+    f("": Any) // warn
+    ^
+error: No warnings can be incurred under -Werror.
+2 warnings
+1 error

--- a/test/files/neg/t12044.scala
+++ b/test/files/neg/t12044.scala
@@ -1,0 +1,56 @@
+//> using options -Werror -Xlint:infer-any
+
+import language.implicitConversions
+
+object retronym {
+  trait Binary[A, B]
+
+  type Unary[A] = Binary[A, A]
+
+  def f[F[A], A](f: F[A]) = ???
+  def test1(u: Unary[Any]) = f(u)
+
+  // reports inference error, cannot unifiy Binary[Any, Any] with ?F[?A]
+  // commented out as we can't have errors in a test for warnings.
+  // def test2(u: Binary[Any, Any]) = f(u)
+
+  def test3 = {
+    implicit def b2u[A, B](b: Binary[A, B]): List[Int] = ???
+    val b: Binary[Any, Any] = null
+    f(b) // inference fails initially, but then we try to coerse the arguments 
+    //
+    // Under -Ytyper-debug, we see:
+    //
+    //    searching for adaptation to pt=Test.Binary[Any,Any] => ?F[?A] (silent: method test3 in Test) implicits disabled
+    // |    |    |    |    5 eligible for pt=Test.Binary[Any,Any] => ?F[?A] at (silent: method test3 in Test) implicits disabled
+    // |    |    |    |    [search #5] considering b2u
+    // |    |    |    |    |-- b2u BYVALmode-EXPRmode-FUNmode-POLYmode (silent: method test3 in Test) implicits disabled
+    // |    |    |    |    |    [adapt] [A, B](b: Test.Binary[A,B])List[Int] adapted to [A, B](b: Test.Binary[A,B])List[Int]
+    // |    |    |    |    |    \-> (b: Test.Binary[A,B])List[Int]
+    // |    |    |    |    solving for (A: ?A, B: ?B)
+    // |    |    |    |    [adapt] [A, B](b: Test.Binary[A,B])List[Int] adapted to [A, B](b: Test.Binary[A,B])List[Int] based on pt Test.Binary[Any,Any] => ?F[?A]
+    // |    |    |    |    [search #5] success inferred value of type Test.Binary[Any,Any] => ?F[?A] is SearchResult(b2u[Any, Any], )
+    //
+    // This leads to inference of ?F=Any, ?A=Nothing (this is kind-correct because Any/Nothing are kind polymorphic)
+    //
+    // When we instantatiate the method type of `f` with this, the formal parameter type is now just Any[Nothing]
+    // which is just Any.
+    //
+    // The provided argument of type `Binary[A, B]` now unifies with this, no implicit coercion required.
+  }
+
+  f[Any, Nothing]("") // explicit type application incurs no warning.
+}
+
+class Bar
+
+object Test extends App {
+  def f[F[_], A](v: F[A]) = v
+  implicit def barToList(b: Bar): List[Int] = List(42)
+  Console.println {
+    f(new Bar) // warn
+  }
+  Console.println {
+    f("": Any) // warn
+  }
+}

